### PR TITLE
Clean up dist_overview.rst

### PR DESCRIPTION
**Pull request by wconstab**
Many overdue updates
* updating the overview to include TP/PP and DTensor/Devicemesh
* removing RPC, DataParallel as they are no longer supported

Fixes #ISSUE_NUMBER

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:74]: can we get rid of the `Single-Machine Model Parallel Best Practices` in this section (inside the second bullet point)?

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:109]: We might want to clean up the https://github.com/pytorch/pytorch/blob/master/torch/distributed/CONTRIBUTING.md doc as a follow up

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:25]: I think these TODOs would likely be rendered in the tutorial page too, so we might want to remove them for now and add them later?

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:31]:
```suggestion
* `DeviceMesh <https://pytorch.org/docs/stable/distributed.html#devicemesh>`__ abstracts the accelerator device communicators into a multi-dimensional array, which manages the underlying ProcessGroups for collective communications in multi-dimensional parallelisms.
```

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:27]: ditto: we probably want to remove the line or we can land after the stable doc link from 2.4?

another option is to use the main doc link instead of stable

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:59]: Shall we unify `Use single-machine multi-GPU` and `Use multi-machine`? IMO the previous doc is to distinguish the old DP and the DDP (DP is single process I think so there were some difference). But now for pytorch distributed we can universally recommend "multi-gpu" DDP

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:51]: nit: gradients are averaged within the data parallel ranks before each optimizer step.

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:34]: do we want to still keep it as "Collective communication (c10d) module"? "C10D" might be a familiar term for us but might not for the audience

**Review by wanchaol — COMMENTED** [review]
The changes looks good to me! Have a few suggestions inlined.

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:25]: oh i was planning to fix before landing. i wanted to ask @awgu what he thinks about doc'ing these two

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:34]: well, any proposal for what to name it then?

File: beginner_source/dist_overview.rst
Comment (svekars) [inline beginner_source/dist_overview.rst:29]: Should those be in backticks as names of modules? if yes, please wrap them, otherwise add to en-wordlist.txt to pass spellcheck. 

Please add `reshard` and `sharded` to en-wordlist.txt.

File: beginner_source/dist_overview.rst
Comment (svekars) [inline beginner_source/dist_overview.rst:31]: Same here. Feels like it should be in backticks (``ProcessGroups ``) - will pass spellcheck after that.

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:29]: DTensor is a class name, not a module name.  Should it be added with backticks too?

If i need to add DTensor to the spelling wordlist, do i need to match the capitalization, or can i just put lowercase dtensor in there and DTensor would also match?

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst]: ok. i didnt see your comments till after i pushed a change to add words to spelling list. ProcessGroup is also a class.  Should i write it as ``ProcessGroup``s?  or where do I put the `s`?   Or do I just put it into the spelling list

Comment (svekars) [reply beginner_source/dist_overview.rst]: `ProcessGroup`s should be okay I think.

File: beginner_source/dist_overview.rst
Comment (svekars) [reply beginner_source/dist_overview.rst:29]: Yes, backticks would work for a class as well. Intersphinx should work too if you'd prefer that.
Both DTensor and dtensor would need to be added. But I'd prefer backticks =) as that's the correct formatting.

File: beginner_source/dist_overview.rst
Comment (svekars) [inline beginner_source/dist_overview.rst:30]: add a blank line above line so that the list renders [correctly](https://docs-preview.pytorch.org/pytorch/tutorials/2913/beginner/dist_overview.html#introduction).

File: beginner_source/dist_overview.rst
Comment (svekars) [reply beginner_source/dist_overview.rst:34]: This whole block looks a bit [weird](https://docs-preview.pytorch.org/pytorch/tutorials/2913/beginner/dist_overview.html) when rendered:
Move `offers both  collective communication APIs (e.g.,` to the same line? Also agree that c10d might be confusing, suggesting to spell it out and give abbreviation in parethesis:

`Collective communication (C10D) module <https://pytorch.org/docs/stable/distributed.html>`__ offers....

File: beginner_source/dist_overview.rst
Comment (svekars) [inline beginner_source/dist_overview.rst:30]: add a blank line before the list so it renders correctly.

File: beginner_source/dist_overview.rst
Comment (svekars) [inline beginner_source/dist_overview.rst:216]: Also, suggesting to add more links in the end to the existing tutorials and other sources that would help users to get started. Could add under the See Also section.

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:34]: Thanks! I decided to reformat this section a bit- added 'headings' for each subsection, and also renamed the c10d link.  lmk what you think about the new version.

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst]: hmm, the backticks worked for DTensor/DeviceMesh, but ProcessGroups rendered badly, like  \``ProcessGroup``s

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst]: hmm, the backticks worked for DTensor/DeviceMesh, but ProcessGroups rendered badly, like  \``ProcessGroup``s

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:216]: yea, that's a good point.  i'm actually not all that familiar with the tutorials content we already have, but i suspect a lot of it is out of date.  I may do a search for any important ones to add now, but i think we can also build that up more over time as we add more tutorials or update old ones.

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:54]: do we still want to brand it as TorchElastic? Or we should just say TorchRun?

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:75]: I wonder if we should split out a model parallelism section given that the beginning of this section only mentioned about "data-parallel training". Or we could augment the first paragraph of this section by adding both Data Parallelism and Model Parallelism, so that it make sense to put this bullet point here.

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:36]: We also have a recipe about DeviceMesh https://pytorch.org/tutorials/recipes/distributed_device_mesh.html?highlight=devicemesh Maybe we should include this recipe in this section? We can put a new sentence in the end of this bullet point: "See a recipe about how to user DeviceMesh in multi-dimensional parallelism https://pytorch.org/tutorials/recipes/distributed_device_mesh.html:

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:75]: Also since we include the model parallelism techniques, could you plug our newly added TP tutorial somewhere? https://pytorch.org/tutorials/intermediate/TP_tutorial.html

**Review by wanchaol — COMMENTED** [review]
I think adding a heading for each section makes things look much nicer! Have a few more suggestions inlined after looked at the previewed doc page.

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:54]: Cc @kurman

Comment (wanchaol) [inline beginner_source/dist_overview.rst:54]: Also it seems we are missing a `__` after the `>`

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:75]: well i renamed the section from 'Data Parallelism' to 'Applying Parallelism To Scale Your Model' to make it more general.

File: beginner_source/dist_overview.rst
Comment (wanchaol) [inline beginner_source/dist_overview.rst:83]: Screenshot 2024-06-12 at 1 10 52 PM

We would need a space before `<`

**Review by wanchaol — APPROVED** [review]
This looks great! Thanks for cleaning up and rewriting our overview documentation!

One small nit and it looks like the CI captured a `scalining` typo

**Comment by GuWei007** [comment]
@wconstab Why torch.distributed.elastic is no longer supported

File: beginner_source/dist_overview.rst
Comment (GuWei007) [reply beginner_source/dist_overview.rst:89]: Why RPC is no longer supported

**Comment by wconstab** [comment]
> @wconstab Why torch.distributed.elastic is no longer supported

I just tried to simplify the 'overview' doc to point to torchrun.  torchrun is the main binary component of torchelastic, and we do support it.

File: beginner_source/dist_overview.rst
Comment (wconstab) [reply beginner_source/dist_overview.rst:89]: RPC has not been actively developed on for a long time, and isn't being used by any of the recently developed distributed components.  I am cleaning up the 'top level' overview to highlight the most active/relevant components.

**Comment by wconstab** [comment]
@svekars can you merge this at your convenience if it looks OK to you?

**Review by svekars — APPROVED** [review]

**Pull request merged**
